storage/innobase/fts/fts0fts.cc

```diff
@@ -6206,3 +5994,464 @@ fts_init_index(
+
+    byte *doc_count= static_cast<byte*>(mem_heap_zalloc(m_heap, 4));
+    mach_write_to_4(doc_count, node->doc_count);
+    dfield_set_data(dtuple_get_nth_field(m_tuple, 5), doc_count, 4);
+  }
+  else
+  {
+    mach_write_to_8(static_cast<byte*>(dtuple_get_nth_field(m_tuple, 1)->data),
+                    node->first_doc_id);
+    mach_write_to_8(static_cast<byte*>(dtuple_get_nth_field(m_tuple, 4)->data),
+                    node->last_doc_id);
+    mach_write_to_4(static_cast<byte*>(dtuple_get_nth_field(m_tuple, 5)->data),
+                    node->doc_count);
+    memset(dtuple_get_nth_field(m_tuple, 2)->data, 0x00, DATA_TRX_ID_LEN);
+    memset(dtuple_get_nth_field(m_tuple, 3)->data, 0x00, DATA_ROLL_PTR_LEN);
+  }
+  dfield_set_data(dtuple_get_nth_field(m_tuple, 6), node->ilist,
+                  node->ilist_size);
+}
+
+bool FTSQueryExecutor::handle_error(dberr_t *err)
+{
+  m_trx->error_state= *err;
+  m_thr->lock_state= QUE_THR_LOCK_ROW;
+  bool lock_wait= row_mysql_handle_errors(err, m_trx, m_thr, m_savept);
+  m_thr->lock_state= QUE_THR_LOCK_NOLOCK;
+  return lock_wait;
+}
+
+dberr_t FTSQueryExecutor::open_table(const char *tbl_name,
+                                     dict_table_t **table,
+                                     bool dict_locked)
+{
+  *table= dict_table_open_on_name(tbl_name, dict_locked,
+                                  DICT_ERR_IGNORE_NONE);
+  if (*table) return DB_SUCCESS;
+  return DB_TABLE_NOT_FOUND;
+}
+
+dberr_t FTSQueryExecutor::open_table(fts_table_t *fts_table,
+                                     dict_table_t **table,
+                                     bool dict_locked)
+{
+  char table_name[MAX_FULL_NAME_LEN];
+  fts_get_table_name(fts_table, table_name, dict_locked);
+  return open_table(table_name, table, dict_locked);
+}
+
+dberr_t FTSQueryExecutor::prepare_for_write(dict_table_t *table)
+{
+  if (!m_thr) create_query_thread();
+  trx_start_if_not_started(m_trx, true);
+  m_thr->graph->trx = m_trx;
+  m_savept= &m_trx->undo_no;
+  dberr_t err= DB_SUCCESS;
+lock_again:
+  err= lock_table(table, nullptr, LOCK_IX, m_thr);
+  if (err && handle_error(&err))
+    goto lock_again;
+  return err;
+}
+
+dberr_t FTSQueryExecutor::prepare_for_read(dict_table_t *table)
+{
+  if (!m_thr) create_query_thread();
+  trx_start_if_not_started(m_trx, true);
+  m_thr->graph->trx = m_trx;
+  m_savept= &m_trx->undo_no;
+  dberr_t err= DB_SUCCESS;
+lock_again:
+  err= lock_table(table, nullptr, LOCK_IS, m_thr);
+  if (err && handle_error(&err))
+    goto lock_again;
+  return err;
+}
+
+dberr_t FTSQueryExecutor::insert_rec(dict_table_t *table)
+{
+  dberr_t err= DB_SUCCESS;
+run_again:
+  err= row_ins_clust_index_entry(dict_table_get_first_index(table),
+                                 m_tuple, m_thr, 0);
+  if (err && handle_error(&err))
+    goto run_again;
+  return err;
+}
+
+dberr_t FTSQueryExecutor::delete_rec(dict_table_t *table)
+{
+  dict_index_t *index= dict_table_get_first_index(table);
```

Comment:
Shouldn’t we assert that there only is one index in the table? Otherwise we’d corrupt secondary indexes.